### PR TITLE
More flexible dashboard widget interface

### PR DESCRIPTION
**Pull request by cprussin**
Partial resolution for #29 .  If you like this @purplg , then I'll follow this up with a PR to add a new widget type for dimmers.

### CODE REVIEW

**Comment by purplg** [comment]
I've only very briefly looked at it for a few minutes. Had some rendering issues with a single group and widget. In particular new lines escapes stayed escaped and everything was on a single line. I might've done something wrong. Maybe my current config bleeding over.

Also I should have mentioned before that you should fork off the Dev branch. I use it as a staging area for master. Shouldn't be too many changes.

I'll take closer look soon!

**Comment by cprussin** [comment]
@purplg  rebased on dev.  I'm not having any issues with newlines, my test config looks like this:

```elisp
(setq hass-dash-layout
      `(group :format ,(propertize "Home Assistant\n%v" 'face 'outline-1)
              (item :format "\n")
              (group :format ,(propertize " Kitchen\n %v" 'face 'outline-2)
                     (hass-dash-toggle :entity-id "light.kitchen_lights")
                     (item :format "\n")
                     (hass-dash-toggle :entity-id "switch.entry_light"))))
```

Can you show me your `hass-dash-layout`?

**Comment by purplg** [comment]
@cprussin Yeah it was my mistake. Had an extra `\` in there I didn't notice that was escaping the `\`.

When I was experimenting with adding multiple entities, I notice that the entities would not automatically stack vertically and instead continued horizontally. I see in your example, you used an `(item "\n")`. Well, it seemed for me that just adding `(item "")` would add a new line and `(item "\n")` would add two. Also I don't see any definition of `item` anywhere so it seems to be some side effect that happens to do almost the right thing.

Anyway, how do you feel about adding a `:layout` or `:orientation` or some such to the group properties? This way each widget doesn't have to figure out its' own placement. For right now it could be horizontal or vertical, but maybe more can be added in the future. Horizontal could be nice for widgets that are just icons.

I've been wanting to add something more robust like this for a while but never came up with a good way to do it. I think this is a great start and I appreciate this effort.

**Comment by cprussin** [comment]
@purplg so `group` and `item` come from https://github.com/emacs-mirror/emacs/blob/master/lisp/wid-edit.el#L2969-L2980 and https://github.com/emacs-mirror/emacs/blob/master/lisp/wid-edit.el#L1858-L1867 respectively in `wid-edit.el`.  This means that we can't modify `group` directly to add `:layout` and `:organization`, but I was planning to add some layout management widgets if you liked the general concept of this PR.  What I had envisioned is something like this:

```elisp
(setq hass-dash-layout
      `(hass-dashboard :title "Home Assistant"
              (hass-dashboard-section :title "Kitchen"
                     (hass-vertical-group
                             (hass-dash-toggle :entity-id "light.kitchen_lights")
                             (hass-dash-toggle :entity-id "switch.entry_light")))))
```

Where `hass-dashboard`, `hass-dashboard-section`, and `hass-vertical-group` (and I'd also make a `hass-horizontal-group`) would be newly defined widgets implemented in terms of `group` and adding some nice formatting.

To be honest, I'm a bit surprised such layout widgets don't already exist somewhere, but I did poke around and I didn't find anything that looked nice and well supported, and it's pretty trivial to roll by hand.

Thoughts?

**Comment by purplg** [comment]
@cprussin 
> @purplg so `group` and `item` come from https://github.com/emacs-mirror/emacs/blob/master/lisp/wid-edit.el#L2969-L2980 and https://github.com/emacs-mirror/emacs/blob/master/lisp/wid-edit.el#L1858-L1867 respectively in `wid-edit.el

Ah! That makes total since. I read a lot about widgets, but I had never come across `wid-edit`. Very convenient. I'll be sure to explore the package more when I have some time.


> This means that we can't modify `group` directly to add `:layout` and `:organization`, but I was planning to add some layout management widgets if you liked the general concept of this PR. What I had envisioned is something like this:
> 
> ```emacs-lisp
> (setq hass-dash-layout
>       `(hass-dashboard :title "Home Assistant"
>               (hass-dashboard-section :title "Kitchen"
>                      (hass-vertical-group
>                              (hass-dash-toggle :entity-id "light.kitchen_lights")
>                              (hass-dash-toggle :entity-id "switch.entry_light")))))
> ```
> 
> Where `hass-dashboard`, `hass-dashboard-section`, and `hass-vertical-group` (and I'd also make a `hass-horizontal-group`) would be newly defined widgets implemented in terms of `group` and adding some nice formatting

Yeah, I do really like the concept. It'll enable the ability to keep relatively simple dashboards or create really complex ones.

> (EDIT: and maybe we could also do some folding for dashboard sections?) (EDIT2: ooo and maybe we could add multiple dashboards that you could bind to different keys or something?).

In regards to your edits, definitely something I'm open to, but maybe a future PR later since this is a pretty big change as it is! Haha.

> To be honest, I'm a bit surprised such layout widgets don't already exist somewhere, but I did poke around and I didn't find anything that looked nice and well supported, and it's pretty trivial to roll by hand.

With the limited functionality I'm envisioning, rolling by hand would probably better anyway. Or, since it doesn't seem to exist, if you decide you wanna really make something fancy with it, maybe make an external package for widget layouts and hass can depend on that :P

**Comment by cprussin** [comment]
@purplg cool, so do you want me to add those layout widgets as part of this PR or shall I do a follow-up PR instead?

**Comment by purplg** [comment]
@cprussin 
I don't have a strong opinion on that.

If the default behavior without the layouts is the same as it is now (vertically stacked), it could be a separate PR. But I wouldn't complain if it was included here.

**Comment by cprussin** [comment]
@purplg OK, I've added `hass-dash-group`.  I've also reformatted the toggle to have a newline, which means the vertical group widget is no longer necessary.  I decided to do this because when I was playing with it, it nearly never makes sense to do things horizontally except when creating abstractions around groups, when you'll be using `:format` anyways and thus the grouping helpers are not useful.

With my change you can now do the following:

```elisp
(setq hass-dash-layout
      `(hass-dash-group :title "Home Assistant"
                        :format "%t\n\n%v"
                        (hass-dash-group :title "Kitchen"
                                         :title-face outline-2
                                         (hass-dash-toggle :entity-id "light.kitchen_lights")
                                         (hass-dash-toggle :entity-id "switch.entry_light"
                                                           :label "Hallway"
                                                           :confirm t))))
```

If you're cool with this and get it merged in, I have another PR queued on top of this that changes `hass-dash-layout` to an alist (now pluralized to `hass-dash-layouts`) and a dashboard selector using `completing-read`.  I'll open that PR once this one lands.

Lastly, I'll put together the dimmer widget as was the original goal, but again I'll wait until you merge this to do so.

Any other changes you'd like to see in this PR?

File: hass-dash.el
Comment (purplg) [inline hass-dash.el:117]: This function doesn't seem to work for me using your layout example but with my entity IDs. Even running it manually, the `hass-tracked-entities` is still set to `nil`. I should contain a list of all entities IDs that a widget state is associated with.

File: hass-dash.el
Comment (purplg) [inline hass-dash.el:112]: I suppose the reason you went with creating a single top-level group is to support different layouts. Totally cool with me. I think it would be nice though if the groups (particularly the top-level one) didn't require a title.

An easy solution for the user is just setting `:format "%v"`, but then they also have to set `:title` to something like an empty string even though it's not used.

Not sure if you have a better solution, but I feel that a title-less group might be a common enough need that it should be explicitly documented at the very least.

**Review by purplg — CHANGES_REQUESTED** [review]
@cprussin 
Made some comments.

This looks really good. Thanks for providing some excellent documentation. :+1:

File: hass-dash.el
Comment (cprussin) [reply hass-dash.el:117]: ahh, this broke as a result of adding the `hass-dash-group` widget, since I'm explicitly looking for `group` widgets to descend into children.  There's a better way to do this that uses the `:children` prop instead, I'll switch to that.

File: hass-dash.el
Comment (cprussin) [reply hass-dash.el:213]: So this is actually broken because we're rebuilding the whole buffer whenever any update comes in for the tracked entities.  But I think I can get away from having to rebuild the buffer now since I can track an alist that tells me which entity ids map to which widgets on the page.  Will do that and update the PR.

File: hass-dash.el
Comment (cprussin) [reply hass-dash.el:112]: For that you can just use the built-in `group` widget (actually the `hass-dash-group` widget is just a wrapper around the `group` widget but adding a title with a face).

**Comment by cprussin** [comment]
@purplg  pushed those fixes, let me know if you think it looks good now, thanks

**Comment by cprussin** [comment]
@purplg I just added one more commit that changes `hass-dash-layout` to expect a list of widgets instead of a single widget which I think is a bit more ergonomic

**Review by purplg — APPROVED** [review]

**Comment by purplg** [comment]
@cprussin Thanks! Merged to dev

**Comment by purplg** [comment]
@cprussin I was expecting you to implement the dimmers you wanted, but that's not necessary for release I guess.

**Comment by cprussin** [comment]
ah sorry @purplg my intent was to experiment with some custom widgets locally in my own config once this backbone stuff gets released.  Once I'm happy with some widgets I'll contribute them back.

**Pull request merged**
